### PR TITLE
Network Resource Manager interface

### PR DESCRIPTION
**Pull request by vyzo**
See https://github.com/libp2p/go-libp2p/issues/1260 for the proposal/design discussion.

### CODE REVIEW

File: network/rcmgr.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+type ResourceScope interface {
+	// ReserveMemory reserves memory/buffer space in the scope.
+	ReserveMemory(size int) error
+	// ReleaseMemory explicitly releases memory previously reserved with ReserveMemory
+	ReleaseMemory(size int)
+
+	// GetBuffer reserves memory and allocates a buffer through the buffer pool.
+	GetBuffer(size int) ([]byte, error)
+	// GrowBuffer grows a previous allocated buffer, reserving the appropriate memory space.
+	GrowBuffer(buf []byte, newsize, copy int) ([]byte, error)
```
Comment (Stebalien) [inline network/rcmgr.go:41]: What's with the `copy` param?

File: network/rcmgr.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	ReleaseBuffer(buf []byte)
+
+	// Stat retrieves current resource usage for the scope.
+	Stat() ScopeStat
+}
+
+// TransactionalScope is a mixin interface for transactional scopes.
+type TransactionalScope interface {
+	// Done ends the transaction scope and releases associated resources.
+	Done()
```
Comment (Stebalien) [inline network/rcmgr.go:52]: Do we have any examples of how we think we might want to use these interfaces? We should start with some toy applications before we spend too much time on these APIs.

File: network/rcmgr.go (hunk 1 above)
Comment (vyzo) [reply network/rcmgr.go:41]: to atomically copy the old buffer while growing/replacing it with a new one;  need to document this.

File: network/rcmgr.go (hunk 2 above)
Comment (vyzo) [reply network/rcmgr.go:52]: The most basic hooks will be inside the transport and muxer implementations.
We need to cater to both them, at the bowels of the stack, and higher level users.

File: network/rcmgr.go (hunk 1 above)
Comment (Stebalien) [reply network/rcmgr.go:41]: I'm still confused. Can't you just copy the old buffer anyways?

Comment (marten-seemann) [reply network/rcmgr.go:41]: Is the idea that stream muxers would use this?
I'm a bit worried about the API here, for example, this doesn't really match the quic-go API.

File: network/rcmgr.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// ProtocolScope is the interface for protocol resource scopes.
+type ProtocolScope interface {
+	ResourceScope
+
+	// Protocols returns the list of protocol IDs constrained by this scope.
+	Protocols() []protocol.ID
+	// IsDefault returns true if this is the default (global) scope.
+	IsDefault() bool
```
Comment (marten-seemann) [inline network/rcmgr.go:62]: What would be a non-default scope?

File: network/rcmgr.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+// PeerScope is the interface for peer resource scopes.
+type PeerScope interface {
+	ResourceScope
+
+	// Peer returns the peer ID for this scope
+	Peer() peer.ID
+
+	// OpenSconnect creates a new connection scope for this peer.
+	OpenConnection(dir Direction, usefd bool) (ConnectionScope, error)
```
Comment (marten-seemann) [inline network/rcmgr.go:76]: This should probably be a `...ConnectionTrait`. WebRTC connections for example might be considered more expensive than a lean TCP / QUIC connection.

File: network/rcmgr.go (hunk 1 above)
Comment (vyzo) [reply network/rcmgr.go:41]: I would like to make it atomically release and give you a new one. It really is a shortcut for Get new, Release old, without oversubscribing memory.

Maybe we dont need it, but of it is an old buffer growing by 1.5x it seems ugood to have it.

File: network/rcmgr.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	// AddConnection adds a previously associated connection to this peer scope.
+	AddConnection(ConnectionScope) error
```
Comment (marten-seemann) [inline network/rcmgr.go:78]: Why is this needed? A connection doesn't change it's peer.

File: network/rcmgr.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	ProtocolScope() ProtocolScope
+}
+
+// ScopeStat is a struct containing resource accounting information.
+type ScopeStat struct {
+	NumPeers   int
+	NumConns   int
+	NumStreams int
+
+	Memory int
```
Comment (marten-seemann) [inline network/rcmgr.go:115]:
```suggestion
	Memory int64
```

**Review by marten-seemann — COMMENTED** [review]
Initial pass.

File: network/rcmgr.go (hunk 3 above)
Comment (vyzo) [reply network/rcmgr.go]: I removed this method, it doesn't make much sense.

File: network/rcmgr.go (hunk 4 above)
Comment (vyzo) [reply network/rcmgr.go:93]: maybe... but that would be reflected in their memory reservations, no?

Let's cross that bridge when we get there.

File: network/rcmgr.go (hunk 5 above)
Comment (vyzo) [reply network/rcmgr.go]: removed this method, (unassociated) connections now have a method to set their peer once.

File: network/rcmgr.go (hunk 6 above)
Comment (vyzo) [reply network/rcmgr.go]: done.

File: network/rcmgr.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	ErrResourceLimitExceeded = errors.New("resource limit exceeded")
+	ErrResourceScopeClosed   = errors.New("resource scope closed")
+)
+
+// ResourceManager is the interface to the network resource management subsystem
+type ResourceManager interface {
+	// GetSystem retrieves the system wide resource scope
+	GetSystem() ResourceScope
+	// GetDMZ retrieves the DMZ resource scope
+	GetDMZ() ResourceScope
```
Comment (marten-seemann) [inline network/rcmgr.go:20]: I still find it super weird to call this DMZ. Can we call it system scope?

File: network/rcmgr.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// PeerScope is the interface for peer resource scopes.
+type PeerScope interface {
+	ResourceScope
+
+	// Peer returns the peer ID for this scope
+	Peer() peer.ID
+
+	// OpenSconnect creates a new connection scope for this peer.
```
Comment (marten-seemann) [inline network/rcmgr.go:87]:
```suggestion
	// OpenConnection creates a new connection scope for this peer.
```

File: network/rcmgr.go (hunk 4 above)
Comment (marten-seemann) [reply network/rcmgr.go:88]: ok, works for me

**Review by marten-seemann — COMMENTED** [review]
Quick review today, more to follow tomorrow.

File: network/rcmgr.go (hunk 7 above)
Comment (vyzo) [reply network/rcmgr.go:20]: Well, the system scope is the global hard limits; the DMZ is for floating resources that have not been associated with the appropriate scope yet (inbound connections, unassociated streams, etc).

But agreed, let's find a better name.

File: network/rcmgr.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+// StreamScope is the interface for stream resource scopes
+type StreamScope interface {
+	ResourceScope
+	TransactionalScope
+
+	// ProtocolScope returns the protocol resource scope associated with this stream.
+	// It returns nil if the stream is not associated with any scope.
+	ProtocolScope() ProtocolScope
+	// SetProtocol sets the protocol for a previously unnegotiated stream
+	SetProtocol(proto protocol.ID) error
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:118]: Yes indeed. It also should error if you call it twice.

File: network/rcmgr.go (hunk 7 above)
Comment (vyzo) [reply network/rcmgr.go:20]: how about we call it `FloatingScope`?

Comment (vyzo) [reply network/rcmgr.go:20]: or `TransientScope`

File: network/rcmgr.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+// PeerScope is the interface for peer resource scopes.
+type PeerScope interface {
+	ResourceScope
+
+	// Peer returns the peer ID for this scope
+	Peer() peer.ID
+
+	// OpenConnection creates a new connection scope for this peer.
+	OpenConnection(dir Direction, useFD bool) (ConnectionScope, error)
```
Comment (marten-seemann) [inline network/rcmgr.go:88]: We've been talking about making it possible to dial a multiaddr without prior knowledge of the peer ID (we'd learn the peer ID from the handshake).

Should we simplify the API here and remove `OpenConnection` from the `PeerScope`? When dialing a connection to a known peer, we'd always go through the transient scope instead (and immediately call `SetPeer` then).

Comment (vyzo) [reply network/rcmgr.go:88]: makes sense, will do.

Comment (vyzo) [reply network/rcmgr.go]: done.

File: network/rcmgr.go (hunk 7 above)
Comment (vyzo) [reply network/rcmgr.go]: renamed to `GetTransient`.

File: network/rcmgr.go (hunk 1 above)
Comment (vyzo) [reply network/rcmgr.go]: removed the ncopy parameter, implementation should unconditionally copy.

File: network/rcmgr.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+	ViewPeer(peer.ID, func(PeerScope) error) error
+
+	// OpenConnection creates a new connection scope not yet associated with any peer; the connection
+	// is scoped at the transient scope.
+	OpenConnection(dir Direction, usefd bool) (ConnectionScope, error)
+
+	// OpenStream creates a new stream scope, initially unnegotiated.
+	// An unnegotiated stream will be initially unattached to any protocol scope
+	// and constrained by the transient scope.
+	OpenStream(p peer.ID, dir Direction) (StreamScope, error)
```
Comment (marten-seemann) [inline network/rcmgr.go:35]: What was the reason this is not in the `PeerScope`?

Comment (vyzo) [reply network/rcmgr.go:35]: consistency, and it also simplifies locking.

File: network/rcmgr.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+	// Stat retrieves current resource usage for the scope.
+	Stat() ScopeStat
+
+	// BeginTxn creates a new transactional scope rooted at this scope
+	BeginTxn() (TransactionalScope, error)
+}
+
+// Buffer is a wrapper for a buffer managed by a resource scope
+type Buffer interface {
```
Comment (marten-seemann) [inline network/rcmgr.go:59]: Should we make this a
```go
type Buffer []byte
```

I wish we could switch out the `append` function that Go uses!

Comment (vyzo) [reply network/rcmgr.go:59]: we can't, we can't hash or compare a slice, which makes it impossible to track  :(:(

File: network/rcmgr.go [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+import (
+	"errors"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+var (
+	ErrResourceLimitExceeded = errors.New("resource limit exceeded")
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:11]: they could I guess -- do we gain anything?

File: network/rcmgr.go [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+	ReleaseMemory(size int)
+
+	// GetBuffer reserves memory and allocates a buffer through the buffer pool.
+	GetBuffer(size int) (Buffer, error)
+
+	// Stat retrieves current resource usage for the scope.
+	Stat() ScopeStat
+
+	// BeginTxn creates a new transactional scope rooted at this scope
+	BeginTxn() (TransactionalScope, error)
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:55]: ok

File: network/rcmgr.go [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+	ServiceScope() ServiceScope
+	// SetService sets the service owning this stream
+	SetService(srv string) error
+
+	// PeerScope returns the peer resource scope associated with this stream.
+	PeerScope() PeerScope
+}
+
+// UserStreamScope is the user view of a StreamScope
+type UserStreamScope interface {
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:136]: Indeed, we could call the other two scopes `*ManagementScope`.

[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:136]: In fact we should also narrow it to only include SetPeer, Stat, and BeginTxn.

File: network/rcmgr.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:55]: done.

File: network/rcmgr.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go]: Renamed the full scopes to *ManagementScope and dropped the User prefix for user scopes. 

I haven't narrowed the interface though, that's something to think about.

File: network/rcmgr.go (hunk 12 above)
Comment (vyzo) [reply network/rcmgr.go]: RIP buffers.

File: network/rcmgr.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (marten-seemann) [reply network/rcmgr.go:11]: Applications could catch this with a general "if it's just a temporary error, try again in a bit" logic.

[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:11]: ok.

[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go]: done.

File: network/errors.go [hunk 16]
```diff
@@ -1,6 +1,17 @@
 package network
 
-import "errors"
+import (
+	"errors"
+	"net"
+)
+
+type TemporaryError string
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/errors.go:8]: sure, but isnt it generally useful?

[the remark this answers is not in the document]
Comment (vyzo) [reply network/errors.go]: unexported for now; if anyone needs it, we can always upcase it later.

File: network/rcmgr.go [hunk 17]
```diff
@@ -0,0 +1,132 @@
+package network
+
+import (
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+// ResourceManager is the interface to the network resource management subsystem
+type ResourceManager interface {
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:9]: of course! We'll work on the docs once we have moved further forward with the integration and testing.

File: network/conn.go [hunk 18]
```diff
@@ -68,3 +66,10 @@ type ConnStat interface {
 	// Stat stores metadata pertaining to this conn.
 	Stat() ConnStats
 }
+
+// ConnScope is the interface that one can mix into a connection interfac to give it a resource
```
Comment (marten-seemann) [inline network/conn.go:70]: Typo: interface.

Comment (vyzo) [reply network/conn.go]: fixed.

File: network/rcmgr.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:76]: updated.

File: network/rcmgr.go [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+	PeerScope() PeerScope
+}
+
+// UserStreamScope is the user view of a StreamScope
+type StreamScope interface {
+	ResourceScope
+
+	// SetService sets the service owning this stream
+	SetService(srv string) error
+}
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go]: The new big phat godoc on the ResourceManager should explain all this.

The Management scope distinction is also more salient now.

File: network/rcmgr.go [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+	ReserveMemory(size int) error
+	// ReleaseMemory explicitly releases memory previously reserved with ReserveMemory
+	ReleaseMemory(size int)
+
+	// Stat retrieves current resource usage for the scope.
+	Stat() ScopeStat
+
+	// BeginTransaction creates a new transactional scope rooted at this scope
+	BeginTransaction() (TransactionalScope, error)
+}
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:130]: done.

File: network/rcmgr.go [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+	// The caller owns the returned scope and is responsible for calling Done in order to signify
+	// the end of th scope's span.
+	OpenStream(p peer.ID, dir Direction) (StreamManagementScope, error)
+
+	// Close closes the resource manager
+	Close() error
+}
+
+// MemoryStatus is an indicator of the current level of available memory for scope reservations.
+type MemoryStatus int
```
Comment (marten-seemann) [inline network/rcmgr.go:119]:
```suggestion
type MemoryStatus uint8
```

Comment (vyzo) [reply network/rcmgr.go:119]: does this change anything? It is not stored anywhere, so it is not using memory and it is just a register in either case.

Comment (marten-seemann) [reply network/rcmgr.go:119]: Not sure, but it can't hurt, can it?

Comment (vyzo) [reply network/rcmgr.go:119]: yeah, I guess not and it might lead to shorter code in x86(-64) because it wont have the prefix byte!

File: network/rcmgr.go [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+	// ReserveMemory reserves memory/buffer space in the scope.
+	ReserveMemory(size int) error
+	// ReleaseMemory explicitly releases memory previously reserved with ReserveMemory
+	ReleaseMemory(size int)
+
+	// GetBuffer reserves memory and allocates a buffer through the buffer pool.
+	GetBuffer(size int) (Buffer, error)
+
+	// Stat retrieves current resource usage for the scope.
+	Stat() ScopeStat
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:154]: yeah observability is very important.

We can add cli commands in ipfs and lotus to inspect resource usage.
So it's a keeper.

File: network/rcmgr.go [hunk 23]
```diff
@@ -0,0 +1,317 @@
...
+// and file  descriptors. These account for both space and time used by
+// the stack, as each resource has a direct effect on the system
+// availability and performance.
+//
+// The modus operandi of the resource manager is to restrict resource usage at the time of
+// reservation. When a component of the stack needs to use a resource, it reserves it in the
+// appropriate scope. The resource manager gates the reservation against the scope applicable
+// limits; if the limit is exceeded, then an error (wrapping ErrResourceLimitExceeded) and it
+// is up the component to act accordingly. At the lower levels of the stack, this will normally
+// signal a filure of some sorts, like failing to opening a stream or a connection, which will
```
Comment (raulk) [inline network/rcmgr.go:38]:
```suggestion
// signal a failure of some sorts, like failing to opening a stream or a connection, which will
```

Comment (raulk) [inline network/rcmgr.go:38]: Did you consider defining a normalized errors for components to wrap when the resource manager denies an action, and this prevents progress?

File: network/rcmgr.go [hunk 24]
```diff
@@ -0,0 +1,317 @@
...
+	ViewTransient(func(ResourceScope) error) error
+
+	// ViewService retrieves a service-specific scope.
+	ViewService(string, func(ServiceScope) error) error
+
+	// ViewProtocol views the resource management scope for a specific protocol.
+	ViewProtocol(protocol.ID, func(ProtocolScope) error) error
+
+	// ViewPeer views the resource management scope for a specific peer.
+	ViewPeer(peer.ID, func(PeerScope) error) error
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:99]: sure, can do.

File: network/rcmgr.go (hunk 23 above)
Comment (vyzo) [reply network/rcmgr.go]: Yes, we have ErrResourceLimitExceeded which implements net.Error as temporary.

I should probably add something here about it.

File: network/rcmgr.go [hunk 25]
```diff
@@ -0,0 +1,317 @@
...
+	ReserveMemory(size int, prio uint8) error
+
+	// ReleaseMemory explicitly releases memory previously reserved with ReserveMemory
+	ReleaseMemory(size int)
+
+	// Stat retrieves current resource usage for the scope.
+	Stat() ScopeStat
+
+	// BeginTransaction creates a new transactional scope rooted at this scope
+	BeginTransaction() (TransactionalScope, error)
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:157]: We currently error if the scope has been closed. But it's good to have room for growth too.

File: network/rcmgr.go [hunk 26]
```diff
@@ -0,0 +1,317 @@
...
+	//
+	// The priority argument indicates the priority of the memory reservation. A reservation
+	// will fail if the available memory is less than (1+prio)/256 of the scope limit, providing
+	// a mechanism to gracefully handle optional reservations that might overload the system.
+	// For instance, a muxer growing a window buffer will use a low priority and only grow the buffer
+	// if there is no memory pressure in the system.
+	//
+	// The are 4 predefined priority levels, Low, Medium, High and Always,
+	// capturing common patterns, but the user is free to use any granularity applicable to his case.
+	ReserveMemory(size int, prio uint8) error
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:148]: sure, will do

File: network/rcmgr.go [hunk 27]
```diff
@@ -0,0 +1,317 @@
...
+	// SetService sets the service owning this stream.
+	SetService(srv string) error
+}
+
+// ScopeStat is a struct containing resource accounting information.
+type ScopeStat struct {
+	NumStreamsInbound  int
+	NumStreamsOutbound int
+	NumConnsInbound    int
+	NumConnsOutbound   int
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:254]: what do you mean here? Relayed connections are just a virtual connection on top of some other connection and don't use fds.

File: transport/transport.go [hunk 28]
```diff
@@ -29,6 +29,7 @@ type CapableConn interface {
 	mux.MuxedConn
 	network.ConnSecurity
 	network.ConnMultiaddrs
+	network.ConnScoper
```
[the remark this answers is not in the document]
Comment (vyzo) [reply transport/transport.go:32]: fixed.

File: network/rcmgr.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:148]: done.

File: network/rcmgr.go [hunk 29]
```diff
@@ -0,0 +1,317 @@
...
+	// Reservation PriorityMedium is a reservation priority that indicates a reservation if the scope
+	// memory utilization is at 60% or less.
+	ReservationPriorityMedium uint8 = 152
+	// ReservationPriorityHigh is a reservation prioirity that indicates a reservation if the scope
+	// memory utilization is at 80% or less.
+	ReservationPriorityHigh uint8 = 203
+	// ReservationPriorityAlways is a reservation priority that indicates a reservation if there is
+	// enough memory, regardless of scope utilization.
+	ReservationPriorityAlways uint8 = 255
+)
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:131]: They are just predefined convenience constats; user is free to ignore and just use a raw number.

File: network/rcmgr.go [hunk 30]
```diff
@@ -0,0 +1,317 @@
...
+	// The transient scope accounts for resources that are in the process of
+	// full establishment.  For instance, a new connection prior to the
+	// handshake does not belong to any peer, but it still needs to be
+	// constrained as this opens an avenue for attacks in transient resource
+	// usage. Similarly, a stream that has not negotiated a protocol yet is
+	// constrained by the transient scope.
+	ViewTransient(func(ResourceScope) error) error
+
+	// ViewService retrieves a service-specific scope.
+	ViewService(string, func(ServiceScope) error) error
```
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:93]: No, the service does not deal with conns at all, it only deals with streams.
Conns are really shared resources, so there isn't much of a point to have them owned by a service.

File: network/rcmgr.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:254]: We do count them however in stat.

File: network/rcmgr.go (hunk 23 above)
Comment (vyzo) [reply network/rcmgr.go]: wait, there is a reference to ErrResourceLimitExceeded already here; it is defined in errors.go

File: network/rcmgr.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (vyzo) [reply network/rcmgr.go:121]: done.

**Review by marten-seemann — APPROVED** [review]
